src/core/include/units/unit.h

```diff
@@ -63,44 +64,92 @@ inline constexpr bool can_be_prefixed = false;
  *       it gets the incomplete child's type with the CRTP idiom.
  */
 template<Magnitude auto M, typename U>
-struct scaled_unit : downcast_base<scaled_unit<M, U>> {
+struct scaled_unit {
   static constexpr UNITS_MSVC_WORKAROUND(Magnitude) auto mag = M;
   using reference = U;
 };
 
-template<Dimension D, Magnitude auto M>
-using downcast_unit = downcast<scaled_unit<M, typename dimension_unit<D>::reference>>;
+// TODO: Remove when P1985 accepted
+namespace detail {
 
-template<Unit U1, Unit U2>
-struct same_unit_reference : is_same<typename U1::reference, typename U2::reference> {};
+template<Magnitude auto M, typename U>
+void to_base_scaled_unit(const volatile scaled_unit<M, U>*);
+
+}  // namespace detail
+
+/**
+ * @brief A concept matching all unit types in the library
+ *
+ * Satisfied by all unit types derived from an specialization of :class:`scaled_unit`.
+ */
+template<typename T>
+concept Unit = requires(T* t) { detail::to_base_scaled_unit(t); };
+
+namespace detail {
+
+template<typename>
+inline constexpr bool is_named = false;
+
+}
+
+template<typename T>
+concept NamedUnit = Unit<T> && detail::is_named<T>;
+
+template<Unit auto U1, Unit auto U2>
+struct same_unit_reference : is_same<typename decltype(U1)::reference, typename decltype(U2)::reference> {};
+
+// TODO add checking for `per` and power elements as well
```

Comment:
```suggestion
```

And similarly in another place.

Reply:
Could it be that you meant a recursive check, and not adding them to the disjunction below?

Reply:
Yes, this is what I meant. Here we will have to check if `pow` and `per` are of expected types as well, so concepts like `pow_of` and `per_of` will be needed.